jni/include/native_engines_stream_support.h

```diff
@@ -119,6 +121,110 @@ class NativeEngineIndexInputMediator {
 
 
 
+/**
+ * This class delegates the provided index output to do IO processing.
+ * In most cases, it is expected that IndexOutputWithBuffer was passed down to this,
+ * which eventually have Lucene's IndexOutput to write bytes.
+ */
+class NativeEngineIndexOutputMediator {
+ public:
+  NativeEngineIndexOutputMediator(JNIUtilInterface *_jni_interface,
+                                  JNIEnv *_env,
+                                  jobject _indexOutput)
+      : jni_interface(_jni_interface),
+        env(_env),
+        indexOutput(_indexOutput),
+        bufferArray((jbyteArray) (_jni_interface->GetObjectField(_env,
+                                                                 _indexOutput,
+                                                                 getBufferFieldId(_jni_interface, _env)))),
+        writeBytesMethod(getWriteBytesMethod(_jni_interface, _env)),
+        bufferLength(jni_interface->GetJavaBytesArrayLength(env, bufferArray)),
+        nextWriteIndex() {
+  }
+
+  void writeBytes(uint8_t *source, size_t nbytes) {
+    auto left = nbytes;
+    while (left > 0) {
+      const auto writeBytes = std::min(bufferLength - nextWriteIndex, left);
+
+      // === Critical Section Start ===
```

Comment:
Can we RAII for critical section part

Reply:
why it's needed?
in the critical section, no methods are throwing exceptions

Reply:
Make sense , if there is no expection thrown we indeed don't need RAII , but can gurantee that this code will not be extended in future and no expection ever be thrown ..I was just trying to be cautious here !!!